src/DotVVM.Framework/Compilation/DotvvmViewCompilationService.cs

```diff
@@ -0,0 +1,202 @@
+        {
+            if (masterPages == null)
+            {
+                lock (loadMasterPagesLocker)
+                {
+                    if (masterPages == null) masterPages = new List<DotHtmlFileInfo>();
+                }
+            }
+        }
+        
+        private bool IsDotvvmPresenter(RouteBase r)
+        {
+            var presenter = r.GetPresenter(dotvvmConfiguration.ServiceProvider);
+            return presenter.GetType().IsInstanceOfType(dotvvmConfiguration.RouteTable.GetDefaultPresenter(dotvvmConfiguration.ServiceProvider));
+        }
+
+        public async Task<bool> CompileAll(bool buildInParallel=true, bool forceRecompile = false)
+        {
+            List<DotHtmlFileInfo> controlsToCompile;
+            List<DotHtmlFileInfo> routesToCompile;
+            if (forceRecompile)
+            {
+                routesToCompile = routes;
+                controlsToCompile = controls;
+            }
+            else
+            {
+                routesToCompile = routes.Where(t => t.Status == CompilationState.None).ToList();
+                controlsToCompile = controls.Where(t => t.Status == CompilationState.None).ToList();
+            }
+            
+            var tempMasterPages = new ConcurrentBag<DotHtmlFileInfo>();
+            var compileTasks = routesToCompile.Select(a => new Task(() => BuildView(a, tempMasterPages))).ToList();
+
+            compileTasks.AddRange(controlsToCompile.Select(a => new Task(() => BuildView(a, tempMasterPages))).ToList());
+
+            await ExecuteCompileTasks(compileTasks,buildInParallel);
+
+            while (tempMasterPages.Count > 0)
+            {
+                masterPages = masterPages.Union(tempMasterPages).Distinct(DotHtmlFileInfo.VirtualPathComparer).ToList();
+
+                //.NET Standard 2.1 - better replace with tempMasterPages.Clear()
+                tempMasterPages = new ConcurrentBag<DotHtmlFileInfo>();
+
+                var masterPagesTasks = masterPages.Select(i => new Task(() => BuildView(i, tempMasterPages))).ToList();
+                await ExecuteCompileTasks(masterPagesTasks, buildInParallel);
+            }
+
+            OrderByErrors();
+            return !GetFilesWithFailedCompilation().Any();
+        }
+
+        private async Task ExecuteCompileTasks(List<Task> compileTasks, bool buildInParallel)
+        {
+            if (buildInParallel)
+            {
+                await Task.WhenAll(compileTasks.ToArray());
+            }
+            else
+            {
+                foreach (var task in compileTasks)
+                {
+                    task.RunSynchronously();
+                }
+            }
+        }
+
+        protected virtual void OrderByErrors()
+        {
+            routes = routes.OrderByDescending(r => r.Status).ToList();
+            masterPages = masterPages.OrderByDescending(mp => mp.Status).ToList();
+            controls = controls.OrderByDescending(c => c.Status).ToList();
```

Comment:
This might be running in parallel :facepalm:

Reply:
Thanks I've completely missed that this (and many other things) would be issue during parallel processing. I've rewritten this service to cope better with parallelism.